### PR TITLE
Support content create and update

### PR DESCRIPTION
**Pull request by soulseen**
support create & update content with github

### CODE REVIEW

**Comment by soulseen** [comment]
improve about https://github.com/drone/go-scm/issues/55

File: scm/content.go [hunk 1]
```diff
@@ -16,10 +19,20 @@ type (
...
+		Message   string        `json:"message,omitempty" description:"The commit message."`
+		Data      []byte        `json:"data,omitempty" description:"The new file content, using Base64 encoding."`
+		SHA       string        `json:"sha,omitempty" description:"Required if you are updating a file. The blob SHA of the file being replaced."`
+		Author    *CommitAuthor `json:"author,omitempty" description:"The author of the file."`
+		Committer *CommitAuthor `json:"committer,omitempty" description:"The person that committed the file."`
+	}
+
+	// CommitAuthor represents the author or committer of a commit. The commit
+	// author may not correspond to a User.
+	CommitAuthor struct {
```
Comment (bradrydzewski) [inline scm/content.go]: remove this structure and use https://github.com/drone/go-scm/blob/master/scm/git.go#L41

File: scm/content.go [hunk 2]
```diff
@@ -16,10 +19,20 @@ type (
...
 	ContentParams struct {
-		Ref     string
-		Branch  string
-		Message string
-		Data    []byte
+		Branch    string        `json:"branch,omitempty" description:"The branch name. Default: the repository’s default branch."`
+		Message   string        `json:"message,omitempty" description:"The commit message."`
+		Data      []byte        `json:"data,omitempty" description:"The new file content, using Base64 encoding."`
+		SHA       string        `json:"sha,omitempty" description:"Required if you are updating a file. The blob SHA of the file being replaced."`
+		Author    *CommitAuthor `json:"author,omitempty" description:"The author of the file."`
```
Comment (bradrydzewski) [inline scm/content.go]: to ensure consistency with the rest of the codebase, these should not be pointers

File: scm/content.go [hunk 3]
```diff
@@ -16,10 +19,20 @@ type (
 	// ContentParams provide parameters for creating and
 	// updating repository content.
 	ContentParams struct {
-		Ref     string
-		Branch  string
-		Message string
-		Data    []byte
+		Branch    string        `json:"branch,omitempty" description:"The branch name. Default: the repository’s default branch."`
```
Comment (bradrydzewski) [inline scm/content.go]: remove `description` tags

File: scm/driver/github/content.go [hunk 4]
```diff
@@ -55,20 +71,22 @@ type content struct {
...
-		Name  string    `json:"name"`
-		Email string    `json:"email"`
-		Date  time.Time `json:"date"`
-	} `json:"committer"`
+	Branch    string
+	Message   string
+	Content   []byte
+	SHA       string
+	Author    *scm.CommitAuthor
+	Committer *scm.CommitAuthor
```
Comment (bradrydzewski) [inline scm/driver/github/content.go]: please revert these changes.  the `contentUpdate` structure should not rely on structures in the `scm` package.

File: scm/content.go (hunk 2 above)
Comment (bradrydzewski) [reply scm/content.go]: replace with https://github.com/drone/go-scm/blob/master/scm/git.go#L41

File: scm/driver/github/content.go (hunk 4 above)
Comment (bradrydzewski) [inline scm/driver/github/content.go]: please revert these changes.  the `contentUpdate` structure should not rely on structures in the `scm` package.

File: scm/driver/github/content.go [hunk 5]
```diff
@@ -55,20 +76,19 @@ type content struct {
...
-	Author  struct {
-		Name  string    `json:"name"`
-		Email string    `json:"email"`
-		Date  time.Time `json:"date"`
-	} `json:"author"`
-	Committer struct {
-		Name  string    `json:"name"`
-		Email string    `json:"email"`
-		Date  time.Time `json:"date"`
-	} `json:"committer"`
```
Comment (ashwilliams1) [inline scm/driver/github/content.go]: why where the `json:...` tags removed?  these are required to properly marshal and unmarshal data to json.  Have you tested your code to make sure it works?

File: scm/driver/github/content.go [hunk 6]
```diff
@@ -55,20 +76,19 @@ type content struct {
...
+type contentCreateUpdate struct {
+	Branch    string
+	Message   string
+	Content   []byte
+	SHA       string
+	Author    CommitAuthor
+	Committer CommitAuthor
+}
+
+type CommitAuthor struct {
```
Comment (ashwilliams1) [inline scm/driver/github/content.go]: this should be private (lowercase)

File: scm/driver/github/content.go [hunk 7]
```diff
@@ -55,20 +76,19 @@ type content struct {
...
+	Date  *time.Time
+	Name  string
+	Email string
```
Comment (ashwilliams1) [inline scm/driver/github/content.go]: missing `json` tags to ensure correct naming.  Also why was time changed to a pointer?  do you want to use `omitempty` instead?

File: scm/driver/github/content.go [hunk 8]
```diff
@@ -8,9 +8,8 @@ import (
 	"context"
 	"encoding/base64"
 	"fmt"
-	"time"
-
 	"github.com/drone/go-scm/scm"
+	"time"
```
Comment (ashwilliams1) [inline scm/driver/github/content.go]: import ordering is incorrect.  please revert this change so that `time` is grouped with other standard library imports.

File: scm/content.go [hunk 9]
```diff
@@ -16,10 +18,11 @@ type (
...
 	// updating repository content.
 	ContentParams struct {
-		Ref     string
-		Branch  string
-		Message string
-		Data    []byte
+		Branch    string
+		Message   string
+		Data      []byte
+		SHA       string
```
Comment (ashwilliams1) [inline scm/content.go]: for consistency please rename `SHA` to `Sha` since we use the latter throughout the codebase.

File: scm/driver/github/content.go [hunk 10]
```diff
@@ -29,11 +28,33 @@ func (s *contentService) Find(ctx context.Context, repo, path, ref string) (*scm
 }
 
 func (s *contentService) Create(ctx context.Context, repo, path string, params *scm.ContentParams) (*scm.Response, error) {
-	return nil, scm.ErrNotSupported
+	endpoint := fmt.Sprintf("repos/%s/contents/%s", repo, path)
+	in := new(contentCreateUpdate)
```
Comment (ashwilliams1) [inline scm/driver/github/content.go]: for consistency with the codebase please create the structure like this:

```
in := &contentCreateUpdate{
  Committer: commitAuthor{
    Name: params.Signature.Name,
    Email: params.Signature.Email
  },
  Author: commitAuthor{
    Name: params.Signature.Name,
    Email: params.Signature.Email
  },
  Message: params.Message,
  ...
}
```

File: scm/driver/github/content.go [hunk 11]
```diff
@@ -29,11 +28,33 @@ func (s *contentService) Find(ctx context.Context, repo, path, ref string) (*scm
...
+	in.Content = params.Data
+	in.SHA = params.SHA
+	res, err := s.client.do(ctx, "PUT", endpoint, in, nil)
+	return res, err
 }
 
 func (s *contentService) Update(ctx context.Context, repo, path string, params *scm.ContentParams) (*scm.Response, error) {
-	return nil, scm.ErrNotSupported
+	endpoint := fmt.Sprintf("repos/%s/contents/%s", repo, path)
+	in := new(contentCreateUpdate)
```
Comment (ashwilliams1) [inline scm/driver/github/content.go]: for consistency with the codebase please create the structure like this:

```
in := &contentCreateUpdate{
  Committer: commitAuthor{
    Name: params.Signature.Name,
    Email: params.Signature.Email
  },
  Author: commitAuthor{
    Name: params.Signature.Name,
    Email: params.Signature.Email
  },
  Message: params.Message,
  ...
}
```

File: scm/content.go [hunk 12]
```diff
@@ -16,10 +18,11 @@ type (
 	// ContentParams provide parameters for creating and
 	// updating repository content.
 	ContentParams struct {
-		Ref     string
```
Comment (ashwilliams1) [inline scm/content.go]: the `Ref` attribute should not be removed. It is a placeholder that is required for other providers.
